lm-cli/lm_cli/auth.py

```diff
@@ -0,0 +1,326 @@
+        token_set = load_tokens_from_cache()
+
+    try:
+        identity_data = validate_token_and_extract_identity(token_set)
+    except ExpiredSignatureError:
+        Abort.require_condition(
+            token_set.refresh_token is not None,
+            "The auth token is expired. Please retrieve a new and log in again.",
+            raise_kwargs=dict(
+                subject="Expired access token",
+                support=True,
+            ),
+        )
+
+        logger.debug("The access token is expired. Attempting to refresh token")
+        refresh_access_token(ctx, token_set)
+        identity_data = validate_token_and_extract_identity(token_set)
+
+    logger.debug(f"Persona created with identity data: {identity_data}")
+
+    save_tokens_to_cache(token_set)
+
+    return Persona(
+        token_set=token_set,
+        identity_data=identity_data,
+    )
+
+
+def refresh_access_token(ctx: LicenseManagerContext, token_set: TokenSet):
+    """
+    Attempt to fetch a new access token given a refresh token in a token_set.
+
+    Sets the access token in-place.
+
+    If refresh fails, notify the user that they need to log in again.
+    """
+    url = f"https://{settings.AUTH0_DOMAIN}/oauth/token"
+    logger.debug(f"Requesting refreshed access token from {url}")
+
+    LicenseManagerCliError.require_condition(
+        ctx.client is not None,
+        "Attempted to refresh with a null client. This should not happen",
+    )
+
+    # Make static type-checkers happy
+    assert ctx.client is not None
+
+    refreshed_token_set = cast(
+        TokenSet,
+        make_request(
+            ctx.client,
+            "/oauth/token",
+            "POST",
+            abort_message="The auth token could not be refreshed. Please try logging in again.",
+            abort_subject="EXPIRED ACCESS TOKEN",
+            support=True,
+            response_model_cls=TokenSet,
+            data=dict(
+                client_id=settings.AUTH0_CLIENT_ID,
+                audience=settings.AUTH0_AUDIENCE,
+                grant_type="refresh_token",
+                refresh_token=token_set.refresh_token,
+            ),
+        ),
+    )
+
+    token_set.access_token = refreshed_token_set.access_token
+
+
+def fetch_auth_tokens(ctx: LicenseManagerContext) -> TokenSet:
+    """
+    Fetch an access token (and possibly a refresh token) from Auth0.
+
+    Prints out a URL for the user to use to authenticate and polls the token endpoint to fetch it when
+    the browser-based process finishes
+    """
+    # Make static type-checkers happy
+    assert ctx.client is not None
+
+    device_code_data = cast(
+        DeviceCodeData,
+        make_request(
+            ctx.client,
+            "/oauth/device/code",
+            "POST",
+            expected_status=200,
+            abort_message="There was a problem retrieving a device verification code from the auth provider",
+            abort_subject="COULD NOT RETRIEVE TOKEN",
```

Comment:
```suggestion
            abort_message="There was a problem retrieving a device verification code from the auth provider.",
            abort_subject="COULD NOT RETRIEVE TOKEN.",
```

Reply:
The `abort_message` is concatenated with more information, so it shouldn't have a trailing period.